File: backend/app/schemas/building.py
"""
Building Schemas
"""
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any
from datetime import datetime
from enum import Enum


class BuildingType(str, Enum):
    """Building type enum"""
    RESIDENTIAL = "residential"
    COMMERCIAL = "commercial"
    INDUSTRIAL = "industrial"
    PUBLIC = "public"


class BuildingBase(BaseModel):
    """Base building schema"""
    name: Optional[str] = Field(None, max_length=255)
    building_type: Optional[BuildingType] = None
    total_height: float = Field(..., gt=0)
    floor_area: Optional[float] = Field(None, gt=0)
    floor_count: Optional[int] = Field(None, gt=0)
    reflective_rate: float = Field(default=0.3, ge=0, le=1)
    address: Optional[str] = None
    district: Optional[str] = None
    city: str = Field(default="未知城市", max_length=100)
    country: str = Field(default="China", max_length=50)


class BuildingCreate(BuildingBase):
    """Building creation schema"""
    footprint: Dict[str, Any] = Field(..., description="GeoJSON Polygon")


class BuildingUpdate(BaseModel):
    """Building update schema"""
    name: Optional[str] = Field(None, max_length=255)
    building_type: Optional[BuildingType] = None
    total_height: Optional[float] = Field(None, gt=0)
    floor_area: Optional[float] = Field(None, gt=0)
    floor_count: Optional[int] = Field(None, gt=0)
    reflective_rate: Optional[float] = Field(None, ge=0, le=1)
    address: Optional[str] = None
    district: Optional[str] = None
    city: Optional[str] = None
    footprint: Optional[Dict[str, Any]] = None


class BuildingResponse(BaseModel):
    """Building response schema"""
    id: str
    name: Optional[str] = None
    building_type: Optional[BuildingType] = None
    footprint: Dict[str, Any]
    total_height: float
    floor_area: Optional[float] = None
    floor_count: Optional[int] = None
    reflective_rate: float
    address: Optional[str] = None
    district: Optional[str] = None
    city: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class BuildingListResponse(BaseModel):
    """Building list response schema"""
    buildings: list[BuildingResponse]
    total: int

File: frontend/src/types/building.ts
import { BuildingType, Coordinate } from './common'

export interface Building {
  id: string
  name?: string
  building_type: BuildingType

  // Geometry
  footprint: GeoJSON.Polygon
  height: number // meters
  floor_area?: number // square meters
  floor_count?: number

  // Optical properties
  reflective_rate?: number // 0-1

  // Metadata
  address?: string
  district?: string
  city?: string
  country?: string

  // Analysis data (optional)
  shadow?: ShadowData
  sunlight_stats?: SunlightStats

  created_at: string
  updated_at: string
}

export interface ShadowData {
  polygon: GeoJSON.Polygon
  area: number
  shadow_length_coefficient?: number
}

export interface SunlightStats {
  total_hours: number
  sunlight_hours: number
  sunlight_rate: number
  avg_daily_hours: number
  peak_hours: number
}

export interface BuildingImportParams {
  file: File
  city?: string
}

export interface BuildingImportResult {
  success_count: number
  failed_count: number
  errors: string[]
}

export interface BuildingQueryParams {
  min_lat: number
  max_lat: number
  min_lng: number
  max_lng: number
  include_shadow?: boolean
  analysis_date?: string
  analysis_hour?: number
}

export interface BuildingListResponse {
  buildings: Building[]
  total: number
}
